core/src/main/scala/org/apache/spark/rpc/akka/AkkaRpcEnv.scala

```diff
@@ -0,0 +1,355 @@
+          case message: Any => {
+            logWarning(s"Unknown message: $message")
+          }
+        }
+
+        override def postStop(): Unit = {
+          unregisterEndpoint(endpoint.self)
+          safelyCall(endpoint) {
+            endpoint.onStop()
+          }
+        }
+
+        }), name = name)
+      endpointRef = new AkkaRpcEndpointRef(defaultAddress, actorRef, conf)
+      endpointRef
+    } finally {
+      latch.countDown()
+    }
+  }
+
+  /**
+   * Run `action` safely to avoid to crash the thread. If any non-fatal exception happens, it will
+   * call `endpoint.onError`. If `endpoint.onError` throws any non-fatal exception, just log it.
+   */
+  private def safelyCall(endpoint: RpcEndpoint)(action: => Unit): Unit = {
+    try {
+      action
+    } catch {
+      case NonFatal(e) => {
+        try {
+          endpoint.onError(e)
+        } catch {
+          case NonFatal(e) => logError(s"Ignore error: ${e.getMessage}", e)
+        }
+      }
+    }
+  }
+
+  private def akkaAddressToRpcAddress(address: Address): RpcAddress = {
+    RpcAddress(address.host.getOrElse(defaultAddress.host),
+      address.port.getOrElse(defaultAddress.port))
+  }
+
+  override def setupDriverEndpointRef(name: String): RpcEndpointRef = {
+    new AkkaRpcEndpointRef(defaultAddress, AkkaUtils.makeDriverRef(name, conf, actorSystem), conf)
+  }
+
+  override def setupEndpointRefByUrl(url: String): RpcEndpointRef = {
+    val timeout = Duration.create(conf.getLong("spark.akka.lookupTimeout", 30), "seconds")
+    val ref = Await.result(actorSystem.actorSelection(url).resolveOne(timeout), timeout)
+    // TODO defaultAddress is wrong
+    new AkkaRpcEndpointRef(defaultAddress, ref, conf)
+  }
+
+  override def setupEndpointRef(
+      systemName: String, address: RpcAddress, endpointName: String): RpcEndpointRef = {
+    setupEndpointRefByUrl(uriOf(systemName, address, endpointName))
+  }
+
+  override def uriOf(systemName: String, address: RpcAddress, endpointName: String): String = {
+    AkkaUtils.address(
+      AkkaUtils.protocol(actorSystem), systemName, address.host, address.port, endpointName)
+  }
+
+
+  override def shutdown(): Unit = {
+    actorSystem.shutdown()
+  }
+
+  override def stop(endpoint: RpcEndpointRef): Unit = {
+    require(endpoint.isInstanceOf[AkkaRpcEndpointRef])
+    actorSystem.stop(endpoint.asInstanceOf[AkkaRpcEndpointRef].actorRef)
+  }
+
+  override def awaitTermination(): Unit = {
+    actorSystem.awaitTermination()
+  }
+
+  override def toString = s"${getClass.getSimpleName}($actorSystem)"
+}
+
+private[spark] object AkkaRpcEnv {
+
+  def apply(config: RpcEnvConfig): RpcEnv = {
+    val (actorSystem, boundPort) = AkkaUtils.createActorSystem(
+        config.name, config.host, config.port, config.conf, config.securityManager)
+    new AkkaRpcEnv(actorSystem, config.conf, boundPort)
+  }
+
+}
+
+private[akka] class AkkaRpcEndpointRef(
+    @transient defaultAddress: RpcAddress,
+    val actorRef: ActorRef,
```

Comment:
ActorRef itself is the name of the class in Akka serving as the handler of an actor instance. I just feel that, to illustrate the functionality of this actor, we need to have a more clear name, 

e.g. we didn't call the actor in  `CoarseGrainedSchedulerBackend` as actorRef, instead, we call it `driverActor`